deps/rabbit/src/rabbit_amqqueue.erl

```diff
@@ -1306,100 +1237,35 @@ is_in_virtual_host(Q, VHostName) ->
 
 -spec list(vhost:name()) -> [amqqueue:amqqueue()].
 list(VHostPath) ->
-    All = list(VHostPath, rabbit_queue),
+    All = rabbit_db_queue:get_all(VHostPath),
     NodesRunning = rabbit_nodes:all_running(),
     lists:filter(fun (Q) ->
                          Pid = amqqueue:get_pid(Q),
                          St = amqqueue:get_state(Q),
                          St =/= stopped orelse lists:member(node(Pid), NodesRunning)
                  end, All).
 
-list(VHostPath, TableName) ->
-    list_with_possible_retry(fun() -> do_list(VHostPath, TableName) end).
-
-%% Not dirty_match_object since that would not be transactional when used in a
-%% tx context
-do_list(VHostPath, TableName) ->
-    mnesia:async_dirty(
-      fun () ->
-              mnesia:match_object(
-                TableName,
-                amqqueue:pattern_match_on_name(rabbit_misc:r(VHostPath, queue)),
-                read)
-      end).
-
-list_with_possible_retry(Fun) ->
-    %% amqqueue migration:
-    %% The `rabbit_queue` or `rabbit_durable_queue` tables
-    %% might be migrated between the time we query the pattern
-    %% (with the `amqqueue` module) and the time we call
-    %% `mnesia:dirty_match_object()`. This would lead to an empty list
-    %% (no object matching the now incorrect pattern), not a Mnesia
-    %% error.
-    %%
-    %% So if the result is an empty list and the version of the
-    %% `amqqueue` record changed in between, we retry the operation.
-    %%
-    %% However, we don't do this if inside a Mnesia transaction: we
-    %% could end up with a live lock between this started transaction
-    %% and the Mnesia table migration which is blocked (but the
-    %% rabbit_feature_flags lock is held).
-    AmqqueueRecordVersion = amqqueue:record_version_to_use(),
-    case Fun() of
-        [] ->
-            case mnesia:is_transaction() of
-                true ->
-                    [];
-                false ->
-                    case amqqueue:record_version_to_use() of
-                        AmqqueueRecordVersion -> [];
-                        _                     -> Fun()
-                    end
-            end;
-        Ret ->
-            Ret
-    end.
-
 -spec list_down(rabbit_types:vhost()) -> [amqqueue:amqqueue()].
 
 list_down(VHostPath) ->
     case rabbit_vhost:exists(VHostPath) of
         false -> [];
         true  ->
             Alive = sets:from_list([amqqueue:get_name(Q) || Q <- list(VHostPath)]),
-            Durable = list(VHostPath, rabbit_durable_queue),
             NodesRunning = rabbit_nodes:all_running(),
-            lists:filter(fun (Q) ->
+            rabbit_db_queue:filter_all_durable(
+              VHostPath, fun (Q) ->
```

Comment:
Ouch, thanks! It's now part of the FilterFun. I had initially implemented `filter_all_durable/1` and `filter_all_durable/2` to pass the VHost, but it's really not needed as it can be part of the FilterFun and we have a single function to maintain. I forgot to change this one before pushing it.